azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -526,3 +526,184 @@ def _handle_pipeline_event(self, event):
+    intervals will eventually become a parameter.
+
+    For each operation that needs a timeout check, this stage will add a timer to
+    the operation.  If the timer elapses, this stage will fail the operation with
+    a PipelineTimeoutError.  The intention is that a higher stage will know what to
+    do with that error and act accordingly (either return the error to the user or
+    retry).
+
+    This stage currently assumes that all timed out operatoins are just "lost".
+    It does not attempt to cancel the operation, as Paho doesn't have a way to
+    cancel an operation, and with QOS=1, sending a pub or sub twice is not
+    catastrophic.
+
+    The long term intention of this stage is that the timeout intervals themselves
+    will eventually come from a retry policy.
+
+    Also, as a long-term plan, the operations that need to be watched for timeout
+    will become an initialization parameter for this stage so that differet
+    instances of this stage can watch for timeouts on different operations.
+    This will be done because we want a lower-level timeout stage which can watch
+    for timeouts at the MQTT level, and we want a higher- level timeout stage which
+    can watch for timeouts at the iothub level.  In this way, an MQTT operation that
+    times out can be retried as an MQTT operation and a higher-level IoTHub operation
+    which times out can be retried as an IoTHub operation (which might necessitate
+    redoing multiple MQTT operations).
+    """
+
+    def __init__(self):
+        super(TimeoutStage, self).__init__()
+        self.timeout_intervals = {
+            pipeline_ops_mqtt.MQTTSubscribeOperation: 10,
+            pipeline_ops_mqtt.MQTTUnsubscribeOperation: 10,
+        }
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _should_watch_for_timeout(self, op):
+        """
+        Return True if this stage needs to put a timer on this operation
+        """
+        return op.__class__ in self.timeout_intervals
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        if self._should_watch_for_timeout(op):
+            # Create a timer to watch for operation timeout on this op and attach it
+            # to the op.
+            self_weakref = weakref.ref(self)
+
+            @pipeline_thread.invoke_on_pipeline_thread_nowait
+            def on_timeout():
+                this = self_weakref()
+                logger.info("{}({}): returning timeout error".format(this.name, op.name))
+                this._complete_op(
+                    op,
+                    transport_exceptions.PipelineTimeoutError(
+                        "operation timed out before protocol client could respond"
+                    ),
+                )
+
+            logger.debug("{}({}): Creating timer".format(self.name, op.name))
+            op.timeout_timer = Timer(self.timeout_intervals[op.__class__], on_timeout)
```

Comment:
>op.timeout_timer [](start = 12, length = 16)

Similarly to the `retry_timer` below, I don't love dynamically applying an attribute to the op object that is not part of the op definition.